### ISSUE TITLE
Some annotations on PDFs still orphaned.

### ISSUE THREAD

**Issue by dwhly** [hypothesis/client#266]
### Steps to reproduce
1. Go here: https://hyp.is/A2DBQoEWEea7_jODeE5fXQ/arxiv.org/pdf/1606.02960.pdf

### Expected behaviour
Annotation should anchor

### Actual behaviour
It does not-- but others on that document, made at the same time, do.

### Browser/system information
Chrome

![screen shot 2017-03-01 at 9 44 22 am](https://cloud.githubusercontent.com/assets/1413074/23473032/b28e9ae8-fe63-11e6-8632-84afad900d41.png)

**Comment by sean-roberts** [comment]
Concerning annotation with quote: "We therefore foundit necessary to pre-train the model using a standard,word-level  cross-entropy  loss  as  described  in  Sec-tion 3"

This is in the PDF. However, it spans multiple pages. In addition, there is a footnote in-between the last part of the quote ("...as described in Sec-") and the last characters of the quote ("tion 3"). And when walking through the code the root.textContent does not include the quote and thus fails to anchor. The root.textContent does not contain the text because it only contains one page of text at a time.

According to the code, we do not support anchoring or selections made cross pages.

In this research, a few things have come up: 
1) We show no error (outside of the dev console) or indication to the user that we don't support this use-case.
2) I have a debugger placed inside of the anchor functions for pdf. It fires multiple times after the page has already been anchored. I don't know if that is a known thing or not. I can imagine that will cause janky performance when scrolling. Though, since I am not very knowledgable with PDF anchoring, @robertknight is this intentional?

**Comment by sean-roberts** [comment]
We can continue the discussion but this is not a bug but rather the lack of a rather complex feature. Closing ticket

**Comment by dwhly** [comment]
Thanks @sean-roberts ... can you take a look at this one on the same document?  https://hyp.is/oU86YoETEea3F2MLPRnZOA/arxiv.org/pdf/1606.02960.pdf

I couldn't find it's selection or prefix/suffix.

**Comment by robertknight** [comment]
> I have a debugger placed inside of the anchor functions for pdf. It fires multiple times after the page has already been anchored. I don't know if that is a known thing or not.

What you are seeing might be related to the way that anchoring interacts with PDF.js' lazy rendering.

AFAIK PDF.js only renders pages that are in or near the viewport. Other pages just have placeholder entries in the DOM. Therefore we are not able to anchor all annotations to DOM elements when the document initially loads. Instead the client fetches just the text of the full document and locates annotations' text positions and quotes within that. After this initial anchoring, we know which page each annotation appears on, but the page in the DOM might not contain anything than a placeholder, in which case anchoring creates a dummy DOM element within that placeholder, selects it and returns that as the Range. This enables the bucket bar to show the location of annotations for pages that are not rendered yet.

As you scroll around, pages that are scrolled into view are rendered and pages scrolled out of view are replaced with placeholders. We detect this with `MutationObserver` and re-trigger anchoring when that happens so that the placeholder anchors can be replaced with ranges that match the actual text that the user sees [1].

> I can imagine that will cause janky performance when scrolling.

It could. I don't think anyone has measured it carefully. `MutationObserver` provides a certain amount of defence against jank by batching changes into a single notification which is delivered asynchronously. But yes, some actual measurements would be great 😀

[1] PDF.js maintains two layers for each page, a visual layer that the user sees which preserves the precise locations of glyphs as specified in the PDF and a closely aligned text layer which facilitates selecting, copying and searching text.

**Comment by judell** [comment]
@klemay, @jeremydean where you both aware of the fact that we do not support selections in PDFs that cross page boundaries?

**Comment by klemay** [comment]
@judell yep - we have an open issue for this: https://github.com/hypothesis/product-backlog/issues/86
